requre/online_replacing.py

```diff
@@ -292,3 +305,117 @@ def _internal(*args, **kwargs):
+    and replay responses on the next run.
+
+    - The matching is based on `url`.
+    - Removes tokens from the url when saving if needed.
+
+    Can be used with or without parenthesis.
+
+    :param _func: can be used to decorate function (with, or without parenthesis).
+    :param response_headers_to_drop: list of header names we don't want to save with response
+                                        (Will be replaced to `None`.)
+    :param storage_file: file for reading and writing data in storage_object
+    """
+
+    response_headers_to_drop = response_headers_to_drop or []
+
+    def decorator_cover(func):
+        return replace_module_match(
+            what="requests.sessions.Session.request",
+            decorate=RequestResponseHandling.decorator(
+                item_list=["method", "url", "data"],
+                map_function_to_item={"url": remove_password_from_url},
+                response_headers_to_drop=response_headers_to_drop,
+            ),
+            storage_file=storage_file,
+        )(func)
+
+    if _func is None:
+        return decorator_cover
+    else:
+        return decorator_cover(_func)
+
+
+@contextmanager
+def recording(
+    what: str,
+    decorate: Optional[Union[List[Callable], Callable]] = None,
+    replace: Optional[Callable] = None,
+    storage_file: Optional[str] = None,
+    storage_keys_strategy=StorageKeysInspectSimple,
+):
+    original_key_strategy = DataMiner().key_stategy_cls
+    DataMiner().key_stategy_cls = storage_keys_strategy
+    original_storage_file = storage_file
+    PersistentObjectStorage().storage_file = storage_file
+    # ensure that directory structure exists already
+    os.makedirs(os.path.dirname(PersistentObjectStorage().storage_file), exist_ok=True)
+    # Store values and their replacements for modules to be able to revert changes back
+    original_module_items = _parse_and_replace_sys_modules(
+        what=what, decorate=decorate, replace=replace
+    )
+    try:
+        yield Cassette()
+    finally:
+        # dump data to storage file
+        PersistentObjectStorage().dump()
+        # revert back changed functions
+        for module_name, item_list in original_module_items.items():
+            setattr(
+                item_list["parent_obj"],
+                item_list["original_obj"].__name__,
+                item_list["original_obj"],
+            )
+        # revert back changed values of singletons, to go to consistent state before test
+        # it is important in this order
+        PersistentObjectStorage().storage_file = original_storage_file
+        DataMiner().key_stategy_cls = original_key_strategy
+
+
+@contextmanager
+def request_recording(
+    response_headers_to_drop: Optional[List[str]] = None, storage_file=None
+):
+    with recording(
+        what="requests.sessions.Session.request",
+        decorate=RequestResponseHandling.decorator(
+            item_list=["method", "url", "data"],
```

Comment:
same here, and maybe question, why it is here for second time? there is function: `record_requests` and `request_recording`

Reply:
One is decorator (`record_requests`) and one one context manager (`request_recording`). I'll try to improve naming and (in the next PR) implement both as one class to be able to use them as class as well as context manager.